### PR TITLE
1796 - Experience Form

### LINKED ISSUE #1796: Assemble entire Experience form
# Description
Assemble a component that includes all sections of the Experience form. Besides the "Addition Information" textarea, all UI components should already be created. They should all be wrapped in a React Hook Form.

Depending on the experience type, the step 1 and data shape needs to be different. Alternatively, there can be 5 different versions of the component, which compose the rest of the form around the correct step 1 component.

Also build a wrapper component around that which handles API operations.

[Wireframe](https://www.figma.com/proto/PrvXEDac0HVp7dF7SLWVFm/Candidate-profile-interfaces---V5.0.0?node-id=2340%3A4352&scaling=min-zoom&page-id=0%3A1&starting-point-node-id=1%3A2)

# Acceptance Criteria
- [ ] Entire form component viewable in Storybook
- [ ] React-hook form can transform data (if necessary) between a shape that comes from the api, and the shape that the form requires, and back (for submission)
- [ ] Skills added/removed to section 2 are reflected in section 3, and vice versa
- [ ] Skill interactions are validated by unit tests
- [ ] Write graphql API operations to get all data necessary for populating the form, and to submit the form
- [ ] Save unsubmitted form in local storage using hook

### PR DESCRIPTION
**Pull request by esizer**
# Work in Progress

This is still being worked on but I'm going to submit this PR as a draft due to the size of it and engage some opinions on how to best handle this. This is is currently being handled by a single form which accepts the "type" of experience as a prop to determine the first section of the form.

This reduces the amount of components needed to be create but increases the complexities of types, hence the abstraction of types for this component into [their own types.ts file](https://github.com/GCTC-NTGC/gc-digital-talent/blob/7e68f9e78db4e368e878d0e9001790a69ff802ba/frontend/talentsearch/src/js/components/experienceForm/types.ts).

## Summary

Creates the Experience Form by combining previously developed components into a singe component that submits the data to the API.

Resolves #1796 

## Tasks

 - [x] Entire form component viewable in Storybook
 - [x] Develop related story
 - [x] Massage data for mutations
 - [x] Write graphql API operations to get all data necessary for populating the form, and to submit the form
   - [x] submit the form
   - [x] populate the form
 - [x] Skills added/removed to section 2 are reflected in section 3, and vice versa
 - [x] Manually test submission
 - [x] Add unit tests
 - [x] Skill interactions are validated by unit tests
 - [x] Save unsubmitted form in local storage using hook

#### Routing Simplicity

This was able to handle in a single component so we can pass the "type" of experience as a route parameter which will be forwarded onto a single component to load the specific form we want. i.e.

```tsx
...
  {
    path: `/:type`,
    action: (context) => {
      const experienceType = context.params.type as ExperienceType;
      return {
        component: <ExperienceFormContainer experienceType={experienceType} />,
      };
    },
  }
...
```

#### Fewer Components

Instead of requiring five separate component with nearly identical functionality, this has combined those into a single component reducing component bloat.

#### Responsibility

The largest negative to this that I can think of, and I'm sure there are many more, is that we are entrusting this single component with too much responsibility. It must handle the submitting data to five different mutations which inflates the complexity of the singular component.

#### Complicated Data Massaging

Due to this component taking on the responsibility of five mutations, the [data massaging gets a little messy.](https://github.com/GCTC-NTGC/gc-digital-talent/blob/7e68f9e78db4e368e878d0e9001790a69ff802ba/frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx#L54-L153)

This could probably be improved and was my first stab at it. Since this is very much still a work in progress and early, If anyone has some time to provide feedback on how this could be improved, that would be appreciated.

#### Many New Types

This is not necessarily a pro or con of this approach since there would be new types created either way, it just inflates the number associated with a single component. Instead of having a few types at the top, closely coupled, they have been moved to their own file due to the need for many different types to handle all the different data scenarios that need to be handled.

### CODE REVIEW

**Comment by petertgiles** [comment]
I'm trying to understand what the options are here.  As I see it:

Option 1 (What you're doing right now) is one big ExperienceForm where the first child is swapped in dynamically:

ExperienceForm
  - (all the logic to run data wrangling and mutations for every form)
  - AwardDetailsForm/EducationExperienceForm/PersonalExperienceForm etc
  - ExperienceSkills
  - TextArea

Option 2 (What you're thinking of) would be to update the individual forms to run themselves:

AwardDetailsForm (x5)
 - (logic to data wrangle and submit mutations for just awards)
 - The form that already exists
 - ExperienceSkills
 - TextArea

Am I understanding this correctly?

Honestly I'm not sure which I prefer.  I can see the pros and cons of each.

Just thinking out loud... would it be helpful if each Experience form component also exported their own functions for formValuesToSubmitData and executeMutation?  Or if there was a file with helper functions in each *ExperienceForm directory so that the functions were close to the component they're supporting?

Or if the big "if type = xyz" structures for  formValuesToSubmitData and mutation functions were abstracted behind a big facade in a separate file?  The big structures would still be there but they wouldn't be polluting the presentation logic.

Sorry, I wish I could be confident about what the right answer is.  Hopefully my rambling helps to at least focus the question or spark an idea.

**Comment by esizer** [comment]
> 



> I'm trying to understand what the options are here. As I see it:
> 
> Option 1 (What you're doing right now) is one big ExperienceForm where the first child is swapped in dynamically:
> 
> ExperienceForm
> 
> * (all the logic to run data wrangling and mutations for every form)
> * AwardDetailsForm/EducationExperienceForm/PersonalExperienceForm etc
> * ExperienceSkills
> * TextArea
> 
> Option 2 (What you're thinking of) would be to update the individual forms to run themselves:
> 
> AwardDetailsForm (x5)
> 
> * (logic to data wrangle and submit mutations for just awards)
> * The form that already exists
> * ExperienceSkills
> * TextArea
> 
> Am I understanding this correctly?
> 
> Honestly I'm not sure which I prefer. I can see the pros and cons of each.
> 
> Just thinking out loud... would it be helpful if each Experience form component also exported their own functions for formValuesToSubmitData and executeMutation? Or if there was a file with helper functions in each *ExperienceForm directory so that the functions were close to the component they're supporting?
> 
> Or if the big "if type = xyz" structures for formValuesToSubmitData and mutation functions were abstracted behind a big facade in a separate file? The big structures would still be there but they wouldn't be polluting the presentation logic.
> 
> Sorry, I wish I could be confident about what the right answer is. Hopefully my rambling helps to at least focus the question or spark an idea.

Yeah, this is definitely pros and cons to each approach and it's not very elegant either way. I'm working on trying to abstract all the messy bits to their own files now to see if that approach is at least cleaner. I should be pushing something up in the next hour or so and hopefully that makes things a little nicer.

**Comment by esizer** [comment]
## Update

So, I have come abstractions in and made it so the form pre-populates when passed an existing experience ID which then also swaps the type of mutation (create vs update).

### Update Experience Route Pattern

The proposed route for updating a form follows a similar pattern to the create form where we could just pass the ID of the experience as well.

#### Example

```tsx
...
  {
    path: `/:type/edit/:id`,
    action: (context) => {
      const experienceType = context.params.type as ExperienceType;
      const id = context.params.id;
      return {
        component: <ExperienceFormContainer experienceType={experienceType} experienceId={id} />,
      };
    },
  }
...
```

## Notes

I am still not 100% happy with the result of what we see here but due to piecing together five separate forms, this might be as clean as we can get it for now without making them all separate from each other. If anyone has suggestions on a better implementation, I'm much more than happy to explore them.

**Comment by esizer** [comment]
> @esizer The bundle size limits are not set in stone! They're there mostly so we don't get really big without noticing. If you have a reasonable reason for growing the bundle you can up the limit yourself in the PR.

Thanks, i just bumped it by 5kb which should allow this to pass.

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.stories.tsx
Comment (petertgiles) [inline frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.stories.tsx:32]: Should this `any` be `ExperienceDetailsSubmissionData`?

**Review by petertgiles — CHANGES_REQUESTED** [review]
Wow, big one!  The refactoring of some of the default and mutation logic out is a big improvement.

I tried running through the story and I think there are a few small issues.

The award title didn't appear in the submit data:
image

I don't think the data from step 4 belongs in the experience.
image
I think it belongs to experience_skills.  But I'm not sure!  Maybe check with James?  :sweat_smile: 
image

Also, one of the acceptance criteria is to save a partially finished form in local storage in case of refresh.  You can reference the spike: https://github.com/GCTC-NTGC/gc-digital-talent/pull/2183 .

Also, is there a way I can do a real submit in a browser?  Is there a URL?

File: frontend/talentsearch/src/js/components/skills/SkillsInDetail/SkillsInDetail.tsx
Comment (esizer) [reply frontend/talentsearch/src/js/components/skills/SkillsInDetail/SkillsInDetail.tsx:108]: Yeah, it was previously just acting directly on a function passed through props. I updated it to pass the ID up the tree with the function so the parent component knows the ID of the item being removed. Previously, the delete function passed in would need to figure the ID out itself.

With this new approach, this component its notifying the parent which skill is being deleted rather than the parent component needing to figure that out.

Comment (petertgiles) [reply frontend/talentsearch/src/js/components/skills/SkillsInDetail/SkillsInDetail.tsx:108]: Oh, it's passing the id now, of course.  Thanks!

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx
Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx:73]: So, this is kind of a weird and it makes sense to some degree. I will be pushing up a commit to ignore this rule for this line but will explain why.

`useEffect` takes two arguments, a function and an array of dependencies. The function will fire every time one of the dependencies change between renders. The any use of a dependency within the function have the potential to change so it is recommended to be added to the array hence the rule. However, when you pass an empty dependency array, this function only ever fires on the first render (mounting) of the component.

For this particular case, this is simply massaging initial data to be used in the form values and setting them and since we only want this to ever run once, and we do not care if the dependency change since we only want it to run once and can safely ignore the rule.

I initially tried to just use default values for `react-hook-form`, but this doesn't seem to trigger an update where we are using `watch` to update components when the skills update.

**Comment by esizer** [comment]
@JamesHuf If you could take a look at [Peter's comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480#pullrequestreview-946089570) about section 4 and provide some input that would be great!

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx
Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx:52]: Resolved with [ignore linting rules](https://github.com/GCTC-NTGC/gc-digital-talent/commit/02e4063150154655b94ab5848437e256f8d0ac69)

**Comment by esizer** [comment]
> The award title didn't appear in the submit data: image

Resolved with [add award title to submission data](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/40665ef83e1da5abdd63c702281d1b2daa458331)

> Also, is there a way I can do a real submit in a browser? Is there a URL?

@petertgiles I had one but removed it since I was no 100% if it was the final intended URL. I have added it back in [add experience form route](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/b1a53cb54ab1338cbddc2e4fa026309a96ae13d2) for testing and will leave it to be updated in the future if needed.

You can access the form at `/talent/skills-and-experiences/:type` where type is the short form of the form type.

 - award
 - community
 - education
 - personal
 - work

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.stories.tsx
Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.stories.tsx]: Resolved with [annotate experience form update callback args](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/bbafac2361357809c10c35c86e0664b2ea146da7)

**Comment by JamesHuf** [comment]
> @JamesHuf If you could take a look at [Peter's comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480#pullrequestreview-946089570) about section 4 and provide some input that would be great!

I'm pretty sure this is right. The details in step 4 are for the experience itself here:
image

While the "skill in detail" boxes in step 3 are used for the experience_skill details here:
image

It looks to me like you're submitting both types of details correctly.

**Comment by esizer** [comment]
> > @JamesHuf If you could take a look at [Peter's comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480#pullrequestreview-946089570) about section 4 and provide some input that would be great!
> 
> I'm pretty sure this is right. The details in step 4 are for the experience itself here: image
> 
> While the "skill in detail" boxes in step 3 are used for the experience_skill details here: image
> 
> It looks to me like you're submitting both types of details correctly.

Thanks for the quick response!

**Comment by JamesHuf** [comment]
I think the confusion comes from the fact that both experiences and experience_skills have an attribute called "details. I wonder if it would be reasonable to rename one of them to prevent confusion in the future.

**Comment by esizer** [comment]
> I think the confusion comes from the fact that both experiences and experience_skills have an attribute called "details". I wonder if it would be reasonable to rename one of them to prevent confusion in the future.

We could maybe rename the top level experience one from `details` -> `description` or use the form label and say `additional_information`?

**Comment by vd1992** [comment]
For testing purposes at the minimum, I think making the ExperienceandSkills component active so someone can add experiences, then test them with this component would be handy. 

I think everything needed for that might already be ready here?
https://github.com/GCTC-NTGC/gc-digital-talent/tree/main/frontend/talentsearch/src/js/components/applicantProfile

**Comment by vd1992** [comment]
getting a warning in the console.log when navigating to the component locally 
image

**Comment by vd1992** [comment]
> For testing purposes at the minimum, I think making the ExperienceandSkills component active could be handy.
> 
> This one right here https://github.com/GCTC-NTGC/gc-digital-talent/tree/main/frontend/talentsearch/src/js/components/applicantProfile
> 
> Generates this, handy as you can see what've you have been creating https://61e099a1bb1465003a73d3ce-oenvmbmaia.chromatic.com/?path=/story/applicantprofile-experienceandskillspage--with-experiences

Enabled in local, passed in the id for "Me"
Created an experience
Voila, seems to submit just fine and something that can be rendered just fine. Amazing work!

image

**Comment by vd1992** [comment]
an oddity, it appended on skills I don't think I selected

image

image

**Comment by esizer** [comment]
> an oddity, it appended on skills I don't think I selected

I could be wrong here, but this is setup to use the following mutation:

```graphql
mutation updateWorkExperience(
  id: "uuid", updateWorkExperience: {
    skills: {
      sync: [
        {
          id: "uuid", 
          details: "lorem"
        }
      ]
    }
) {
  id
}
```

This was based off the assumption this aligned with the [Laravel way of syncing relationships](https://laravel.com/docs/9.x/eloquent-relationships#syncing-associations).  Do we have anyone who can confirm?

**Comment by esizer** [comment]
> getting a warning in the console.log when navigating to the component locally

This was related to a `<form>` element being in the skills search bar. It was removed in [remove form from search bar](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/e9b37e918979ba279a1d7faa9880ff7654d9dd9b).

**Review by petertgiles — CHANGES_REQUESTED** [review]
Nice, the nuts and bolts all work properly as far as I can tell. :muscle: 

Could you check word limit from the wireframe?  I think Yoni set it artificially low for the storybook story  to be able to easily test it.
image

The forms now save the data locally!  :tada: But don't forget to clear that storage on successful form submission.  It seems like you have coded for it but it's not working for me.  I get leftover data repopulated even after submitting the form.

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx
Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx:75]: @vd1992 and I found when working on another PR that you need to pass initial data into the `watch` function separately. But you can do any data massaging once and pass the same data into `watch` and `defaultValues`.

**Comment by esizer** [comment]
> Nice, the nuts and bolts all work properly as far as I can tell. 💪
> 
> Could you check word limit from the wireframe? I think Yoni set it artificially low for the storybook story to be able to easily test it. image
> 
> The forms now save the data locally! 🎉 But don't forget to clear that storage on successful form submission. It seems like you have coded for it but it's not working for me. I get leftover data repopulated even after submitting the form.

I have updated the number of words to 160 (roughly a medium sized paragraph) in [increase max words for skill details](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/1598958e51b86bbd84c800172e417069c9bcda2a)

As far as clearing local storage, this one was a little odd. It seems as though `react-hook-form` does not reset the form values on submission which makes sense but it also happens to trigger all `watch` values to update which results in the `useEffect` hook to fire again and set the local storage back to the previous value.

To avoid this, I moved the reset function up a component and put it after page navigation in [move local storage reset to after navigation](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/a1da745d7c78f656e54f0cbd2decd35201283537). This is not ideal but I tried many different alternatives (such as clearing form on submission) but none of them seemed to work quite right. This could also be my lack of experience with `react-hook-form` and if anyone has some ideas on how to do this better, that would be awesome!

**Comment by esizer** [comment]
> Sorry, I know I'm late to this conversation, but I really want to weight in on the question of how to structure 5 similar forms:
> 
> > Option 1 (What you're doing right now) is one big ExperienceForm where the first child is swapped in dynamically:
> > ExperienceForm
> > (all the logic to run data wrangling and mutations for every form)
> > AwardDetailsForm/EducationExperienceForm/PersonalExperienceForm etc
> > ExperienceSkills
> > TextArea
> > Option 2 (What you're thinking of) would be to update the individual forms to run themselves:
> > AwardDetailsForm (x5)
> > (logic to data wrangle and submit mutations for just awards)
> > The form that already exists
> > ExperienceSkills
> > TextArea
> 
> I highly recommend the Option 2! I think React does too. (I think of this as the [composition vs inheritance](https://reactjs.org/docs/composition-vs-inheritance.html) dichotomy, even if inheritance isn't quite the right word here.)
> 
> In my experience, the second style has led to components which are easier to refactor and reuse. In fact, I've often found myself refactoring something from Option 1-style to Option 2-style as soon as I wanted to reuse part of it. I'll try to quickly put together an example of how I would do it here.

Okay, we may need to come back and refactor since it is almost complete using the first options philosophy. I may suggest we attempt to roll the separate form sections into the larger parent component since it seems like it may be an unnecessary abstraction. This type of refactor would result in 15 different components where 10 of those essentially do the exact same thing outside of render the first 4-6 fields differently.

Would we be open to merging this as is and creating a new issue to refactor at a later date, or should I continue on with the refactor and push it to the next sprint since we are essentially at the end of this one? @tristan-orourke 

## Components Required for Separate Forms

1. {award | community | education | personal | work}DetailsForm
2. {award | community | education | personal | work}ExperienceForm
3. {award | community | education | personal | work}DetailsExperienceFormWrapper

**Comment by tristan-orourke** [comment]
> Okay, we may need to come back and refactor since it is almost complete using the first options philosophy. I may suggest we attempt to roll the separate form sections into the larger parent component since it seems like it may be an unnecessary abstraction. This type of refactor would result in 15 different components where 10 of those essentially do the exact same thing outside of render the first 4-6 fields differently.
> 
> Would we be open to merging this as is and creating a new issue to refactor at a later date, or should I continue on with the refactor and push it to the next sprint since we are essentially at the end of this one? @tristan-orourke

Go ahead and merge as is. As much as I prefer more components with narrower scope, this style is perfectly functional, and you've done a great job with it.

**Comment by esizer** [comment]
> > Okay, we may need to come back and refactor since it is almost complete using the first options philosophy. I may suggest we attempt to roll the separate form sections into the larger parent component since it seems like it may be an unnecessary abstraction. This type of refactor would result in 15 different components where 10 of those essentially do the exact same thing outside of render the first 4-6 fields differently.
> > Would we be open to merging this as is and creating a new issue to refactor at a later date, or should I continue on with the refactor and push it to the next sprint since we are essentially at the end of this one? @tristan-orourke
> 
> Go ahead and merge as is. As much as I prefer more components with narrower scope, this style is perfectly functional, and you've done a great job with it.

Sounds good, maybe we can just pop something in the backlog to come back and refactor at a later date and I will keep this in mind for future PRs.

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx
Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx]: > @vd1992 and I found when working on another PR that you need to pass initial data into the `watch` function separately. But you can do any data massaging once and pass the same data into `watch` and `defaultValues`.

Wow! Thanks for the tip, still new with `react-hook-form` so I was unaware of this. I have updated it with [simplify watch default values](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/113d91412727477e69b74b03dad42c68f26c868b) and the whole thing is much cleaner now.

**Comment by petertgiles** [comment]
I dunno, it's still saving for me.  Am I missing something?

**Comment by esizer** [comment]
> I dunno, it's still saving for me. Am I missing something? Peek 2022-04-20 16-20 ?

Okay, it was working for me in chrome at least. I moved the local storage bit up above the navigation. Are you able to try again? I tried in a few browsers and it is working for me in each. 🤔

**Comment by vd1992** [comment]
- pulled latest and built 
- filled out form
- successfully submitted it
- navigated to it again
- form is not pre-filled

was not able to repeat the bug
using firefox

**Comment by vd1992** [comment]
bug
cannot submit work experience if "currently active in this role" is selected
receive this
image

**Comment by esizer** [comment]
> fix endDate param for experience submissions

Thanks, good catch. Looks like a `null` value was being passed rather than setting `endDate` to `undefined` which meant it was not a valid graphql mutation.

Resolved with [fix endDate param for experience submissions](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/40c09ca9d7547fa70a59c6ac5cab0a11b5f973b3)

File: frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx
Comment (tristan-orourke) [inline frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Some thoughts about this form caching... 🤔 

It looks liked the cached values will override the defaultValues as long as the key exists in local storage. I can't test it right now, but it sounds like if I work on one Experience, and then I try to edit a second Experience, I'll be stuck looking at the previous one.

One way around this would be to use more storage keys. Maybe ``key = `ts-createExperience-${experience ? experience.id : experienceType}`;``

One way around this could be

Comment (vd1992) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Yeah, if you start one form and navigate to another without submitting, it carries values over

Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: > Some thoughts about this form caching... 🤔
> 
> It looks liked the cached values will override the defaultValues as long as the key exists in local storage. I can't test it right now, but it sounds like if I work on one Experience, and then I try to edit a second Experience, I'll be stuck looking at the previous one.
> 
> One way around this would be to use more storage keys. Maybe `` key = `ts-createExperience-${experience ? experience.id : experienceType}`; `` Then editing each existing experience creates its own cache, and Create New has its own for each different type.
> 
> This would create a lot of keys over time, and local storage hangs around forever. Maybe we should use [session storage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) instead for this. That would protect your data if you navigated away accidentally, but not if you closed the tab. But it also stops old data from accruing.
> 
> Another thought - maybe we clear the cache (at this key) when the form is submitted?

A lot of good ideas in this. So right now we are *kind of* clearing it when the form is submitted, however our hook does not allow to actually delete the item, just set it to a null value. We could look into exporting another function from the hook that allows for [removing an item](https://developer.mozilla.org/en-US/docs/Web/API/Storage/removeItem) as well?

As far a session storage goes, is it *immediately* cleared when a tab is closed? I have a bad habit of closing tabs on accident all the time and assume I'm not the only one. If it deletes immediately I would recommend sticking with local storage and just delete the item once the form is submitted.

Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Just thought of a different problem with using unique keys per experience. What if I edit an experience on one computer, which saves it in localStorage. Then I edit it again on another device. The first device will show the old cached data! My first thought is this would be another good reason to use session storage.

Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Okay, so I'll make a `useSessionStorage` hook and update the form to use that instead.

Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Spun off into issue https://github.com/GCTC-NTGC/gc-digital-talent/issues/2530

Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: > Spun off into issue #2530

Alright, so are we good to merge this then?

Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Yep @esizer, I don't this discussion to block this PR

Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: > As far a session storage goes, is it immediately cleared when a tab is closed? I have a bad habit of closing tabs on accident all the time and assume I'm not the only one. If it deletes immediately I would recommend sticking with local storage and just delete the item once the form is submitted.

Since we're using our own hook, we could use localStorage but build in some logic which includes an expiration time. eg if the hook loads data which is more than an hour old it discards it. That might get around most of these concerns.

Comment (tristan-orourke) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: If we wanted to be REALLY clever we could start including `update_at` timestamps in our api! Maybe that's going too far lol.

Comment (esizer) [reply frontend/talentsearch/src/js/components/experienceForm/ExperienceForm.tsx:67]: Sounds good, I like both ideas. Maybe we should add those into the new issue if you haven't already. 

I'll work on resolving the merge conflict then merge shortly.

**Pull request merged**
